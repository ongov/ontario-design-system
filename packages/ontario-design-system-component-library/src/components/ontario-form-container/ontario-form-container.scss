@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/spacing.variables' as spacing;

// Mixin for generating individual form container gap classes
@mixin ontario-form-container-gap($modifier, $spacing-variable) {
	.ontario-form-container--gap-#{$modifier},
	.ontario-form-container--gap-#{$modifier} form {
		::slotted(ontario-input:not(:last-child)),
		::slotted(ontario-checkboxes:not(:last-child)),
		::slotted(ontario-dropdown-list:not(:last-child)),
		::slotted(ontario-textarea:not(:last-child)),
		::slotted(ontario-radio-buttons:not(:last-child)),
		::slotted(ontario-button:not(:last-child)),
		::slotted(ontario-date-input:not(:last-child)) {
			margin-bottom: (#{$spacing-variable});
			display: block;
		}
	}
}

// Mixin for generating nested form container gap classes
@mixin ontario-nested-form-container-gap($parent-modifier, $parent-spacing, $child-spacings) {
	.ontario-form-container--gap-#{$parent-modifier},
	.ontario-form-container--gap-#{$parent-modifier} form {
		@each $child-spacing in $child-spacings {
			$child-modifier: nth($child-spacing, 1);
			$child-spacing-variable: nth($child-spacing, 2);

			.ontario-form-container--gap-#{$child-modifier},
			.ontario-form-container--gap-#{$child-modifier} form {
				::slotted(ontario-input:not(:last-child)),
				::slotted(ontario-checkboxes:not(:last-child)),
				::slotted(ontario-dropdown-list:not(:last-child)),
				::slotted(ontario-textarea:not(:last-child)),
				::slotted(ontario-radio-buttons:not(:last-child)),
				::slotted(ontario-button:not(:last-child)),
				::slotted(ontario-date-input:not(:last-child)) {
					margin-bottom: $child-spacing-variable;
					display: block;
				}

				// Add extra spacing for the last child
				::slotted(ontario-input:last-child),
				::slotted(ontario-checkboxes:last-child),
				::slotted(ontario-dropdown-list:last-child),
				::slotted(ontario-textarea:last-child),
				::slotted(ontario-radio-buttons:last-child),
				::slotted(ontario-button:last-child),
				::slotted(ontario-date-input:last-child) {
					margin-bottom: $child-spacing-variable + $parent-spacing;
					display: block;
				}
			}
		}
	}
}

// Gap definitions
$gap-definitions: (16, spacing.$spacing-4), (40, spacing.$spacing-7);

// Generate styles for each gap setting
@each $gap-definition in $gap-definitions {
	$gap-modifier: nth($gap-definition, 1);
	$gap-variable: nth($gap-definition, 2);

	// Generate individual form container gap classes
	@include ontario-form-container-gap($gap-modifier, $gap-variable);

	// Generate nested form container gap classes
	$nested-gaps: ();
	@each $nested-gap in $gap-definitions {
		@if $gap-modifier != nth($nested-gap, 1) {
			$nested-gaps: append($nested-gaps, $nested-gap);
		}
	}

	@include ontario-nested-form-container-gap($gap-modifier, $gap-variable, $nested-gaps);
}
