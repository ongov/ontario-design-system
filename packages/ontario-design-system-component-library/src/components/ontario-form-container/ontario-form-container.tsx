import { Component, Element, h, Prop } from '@stencil/core';

@Component({
	tag: 'ontario-form-container',
	styleUrl: 'ontario-form-container.scss',
	shadow: true,
})
export class OntarioFormContainer {
	@Element() element: HTMLElement;

	/**
	 * Defines the gap (bottom margin) between slotted form elements.
	 * If no gap prop is provided, it will default to 'default'.
	 */
	@Prop() gap: 'default' | 'condensed' = 'default';

	/**
	 * Returns the pixel value for the selected gap option.
	 */
	private getGapValue() {
		const gapOptions = {
			default: '40',
			condensed: '16',
		};

		return gapOptions[this.gap] || gapOptions.default;
	}

	render() {
		return (
			<div class={`ontario-form-container--gap-${this.getGapValue()}`}>
				<slot />
			</div>
		);
	}
}
