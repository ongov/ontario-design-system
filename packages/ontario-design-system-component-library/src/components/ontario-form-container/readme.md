import { OntarioFormContainer, OntarioFieldset, OntarioInput, OntarioDropdownList, OntarioTextarea, OntarioButton } from '@ongov/ontario-design-system-component-library-react';

# ontario-form-container

The `ontario-form-container` component defines vertical spacing (referred to as _gap_) between child form elements within a container. It ensures consistent and accessible spacing between form elements, aligning with design tokens from the Ontario Design System.

## Usage Guidance

### When to use this component

Use the `ontario-form-container` component when:

- You need to apply consistent vertical spacing between form elements
- A clear visual hierarchy is required within a form
- Different elements require varying spacing levels (`default` or `condensed`)

### When not to use this component

Avoid using this component if:

- You need horizontal spacing between elements
- The spacing is purely decorative

## Technical Specifications

The component uses a single default slot to render child elements. It is designed to be used with the following form components: `ontario-fieldset`, `ontario-input`, `ontario-textarea`, `ontario-checkboxes`, `ontario-dropdown-list`, `ontario-radio-buttons`, `ontario-date-input` and `ontario-button`.

By default, the above components do not have built-in margins to separate them from one another. The ontario form container component should be used for maintaining readability and creating visual hierarchy.

**Spacing options:**

- **Default** (40px) corresponds to $spacing-7.
- **Condensed** (16px) corresponds to $spacing-4.

**Styling:**

- Applies `margin-bottom` values to all child elements except the last one
- Supports nested usage, allowing smaller spacing within larger spacing containers. Nested spacing should be used when elements are closely linked and require reduced spacing

## Examples

### Default gap example

No `gap` prop is required when using the default gap option.

```html
<form>
	<ontario-form-container>
		<ontario-fieldset legend-size="large" legend="What is your delivery address?">
			<ontario-input
				id="delivery-address-form-line-1"
				caption="Address line 1"
				required
				name="delivery-address-form-line-1"
				hint-text="Street and number or P.O. box."
				enable-live-validation
				required-validation-message="Please enter your address"
			></ontario-input>
			<ontario-input
				caption="Address line 2"
				name="delivery-address-form-line-2"
				hint-text="Apartment, suite, unit, building, etc."
			></ontario-input>
			<ontario-dropdown-list
				caption="Province/territory"
				name="delivery-address-form-province-selection-form"
				element-id="delivery-address-form-province-territory"
				required
				is-empty-start-option="true"
				options='[
					{ "value": "ON", "label": "Ontario" },
					{ "value": "QC", "label": "Quebec" },
					{ "value": "SK", "label": "Saskatchewan" }
				]'
			></ontario-dropdown-list>
		</ontario-fieldset>
	</ontario-form-container>
</form>
```

<form>
	<OntarioFormContainer>
		<OntarioFieldset legendSize="large" legend="What is your delivery address?">
			<OntarioInput
				id="delivery-address-form-line-1"
				caption="Address line 1"
				required
				name="delivery-address-form-line-1"
				hintText="Street and number or P.O. box."
				enableLiveValidation
				requiredValidationMessage="Please enter your address"
			>
			</OntarioInput
				caption="Address line 2"
				name="delivery-address-form-line-2"
				hintText="Apartment, suite, unit, building, etc."
			>
			</OntarioInput>
			<OntarioDropdownList
				caption="Province/territory"
				name="delivery-address-form-province-selection-form"
				elementId="delivery-address-form-province-territory"
				required
				isEmptyStartOption="true"
				options='{[
					{ "value": "ON", "label": "Ontario" },
					{ "value": "QC", "label": "Quebec" },
					{ "value": "SK", "label": "Saskatchewan" }
				]}'
			>
			</OntarioDropdownList>
		</OntarioFieldset>
	</OntarioFormContainer>
</form>

### Mixed gap example

This example uses both default and condensed gaps.

```html
<form>
	<ontario-form-container>
		<ontario-fieldset legend-size="large" legend="What is your delivery address?">
			<ontario-input
				name="email"
				input-width="20-char-width"
				caption='{
                    "captionText": "Email",
                    "captionType": "large"
                }'
				required="true"
			></ontario-input>
			<ontario-form-container gap="condensed">
				<ontario-textarea
					name="comments"
					caption='{
                        "captionText": "Comments",
                        "captionType": "large"
                    }'
					required="true"
				></ontario-textarea>
				<ontario-button type="secondary">✅ Share my feedback</ontario-button>
			</ontario-form-container>
		</ontario-fieldset>
	</ontario-form-container>
</form>
```

<form>
	<OntarioFormContainer>
		<OntarioFieldset legendSize="large" legend="What is your delivery address?">
			<OntarioInput
				name="email"
				inputWidth="20-char-width"
				caption='{
                    "captionText": "Email",
                    "captionType": "large"
                }'
				required="true"
			></OntarioInput>
			<OntarioFormContainer gap="condensed">
				<OntarioTextarea
					name="comments"
					caption='{
                        "captionText": "Comments",
                        "captionType": "large"
                    }'
					required="true"
				></OntarioTextarea>
				<OntarioButton type="secondary">✅ Share my feedback</OntarioButton>
			</OntarioFormContainer>
		</OntarioFieldset>
	</OntarioFormContainer>
</form>

## Best Practices

- Use the **default gap** spacing for standard forms
- Use the **condensed gap** spacing when form elements are closely related but of different types
- Limit the number of nested spacing levels to avoid complex and unreadable layouts

## Accessibility considerations

- **Readability and Usability**: Ensure that spacing choices support clear and structured layouts
- **Focus Order**: Spacing should not disrupt the expected focus order when navigating with a keyboard or screen reader
- **Visual Hierarchy**: Appropriate spacing improves visual hierarchy, reducing cognitive load
- **Semantic Markup**: When using condensed spacing sections, consider using semantic elements to group items together, such as the `<ontario-fieldset>`

## Configuration

Once the component package has been installed (see the Ontario Design System Component Library for installation instructions), the `ontario-form-container` component can be added directly into the project's code and customized using the `gap` property.

<!-- Auto Generated Below -->

## Properties

| Property | Attribute | Description                                                                                                              | Type                       | Default     |
| -------- | --------- | ------------------------------------------------------------------------------------------------------------------------ | -------------------------- | ----------- |
| `gap`    | `gap`     | Defines the gap (bottom margin) between slotted form elements. If no gap prop is provided, it will default to 'default'. | `"condensed" \| "default"` | `'default'` |

---

_Built with [StencilJS](https://stenciljs.com/)_
