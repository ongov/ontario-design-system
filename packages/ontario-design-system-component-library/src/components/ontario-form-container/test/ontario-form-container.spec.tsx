import { newSpecPage } from '@stencil/core/testing';
import { OntarioFormContainer } from '../ontario-form-container';

describe('ontario-form-container', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioFormContainer],
				html: `<ontario-form-container></ontario-form-container>`,
			});
			expect(page.root).toMatchSnapshot();
		});
	});

	it('should render with default gap if no gap prop is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioFormContainer],
			html: `<ontario-form-container></ontario-form-container>`,
		});

		expect(page.root?.shadowRoot?.querySelector('div')).toHaveClass('ontario-form-container--gap-40');
	});

	it('should render with condensed gap if the `condensed` spacing gap is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioFormContainer],
			html: `<ontario-form-container gap="condensed"></ontario-form-container>`,
		});

		expect(page.root?.shadowRoot?.querySelector('div')).toHaveClass('ontario-form-container--gap-16');
	});
});
