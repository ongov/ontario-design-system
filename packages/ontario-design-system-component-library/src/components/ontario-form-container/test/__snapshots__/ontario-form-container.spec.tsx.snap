// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-form-container snapshot should render the expected html 1`] = `
<ontario-form-container>
  <template shadowrootmode="open">
    <div class="ontario-form-container--gap-40">
      <slot></slot>
    </div>
  </template>
  <slot></slot>
</ontario-form-container>
`;
